### PR TITLE
[filter/snpe] Make the order of output tensors be consistent with the user given option @open sesame 4/28 09:30

### PR DESCRIPTION
**Pull request by anyj0527**
- For multiple output models, let user set output tensors instead of
  output layers
- And make the order of input / output tensors be consistent.

Updated:
- Some snpe models have "0" dim to represent resizable dim.
- Add a custom option "MaxResizableDim" to deal with this case.

Signed-off-by: Yongjoo Ahn <yongjoo1.ahn@samsung.com>

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 1]
```diff
@@ -221,17 +220,15 @@ snpe_subplugin::runtimeToString (zdl::DlSystem::Runtime_t runtime)
...
 bool
-snpe_subplugin::set_output_layer_names (const GstTensorsInfo *info)
+snpe_subplugin::set_output_tensor_names (const GstTensorsInfo *info)
 {
   for (unsigned int i = 0; i < info->num_tensors; ++i) {
     if (info->info[i].name == nullptr) {
       /* failed */
-      nns_loge ("Given layer name with index %u is invalid.", i);
-      output_layer_names.clear ();
+      nns_loge ("Given tensor name with index %u is invalid.", i);
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:228]: ```s/tensor/output tensor/```
```s/invalid/invalid, it is null/```
?

And.. do  you need to check if ```info->info[i].name[0] == '\0'``` as well?

Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:228]: If it is failed at ```i > 0```, don't you need to free elements of output_tensor_names_list ?

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 2]
```diff
@@ -221,17 +220,15 @@ snpe_subplugin::runtimeToString (zdl::DlSystem::Runtime_t runtime)
  * @brief Internal method to get names of output layers from tensors information.
  */
 bool
-snpe_subplugin::set_output_layer_names (const GstTensorsInfo *info)
+snpe_subplugin::set_output_tensor_names (const GstTensorsInfo *info)
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:223]: do you need to initialize (clean up) ```output_tensor_names_list``` before the loop?

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc (hunk 1 above)
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc]: > `s/tensor/output tensor/` `s/invalid/invalid, it is null/` ?
> 
> And.. do you need to check if `info->info[i].name[0] == '\0'` as well?

Updated, Thanks!



> If it is failed at `i > 0`, don't you need to free elements of output_tensor_names_list ?

The SNPE SDK does not provide API for freeing elements or cleaning. I updated the `cleanup ()` function to reset the list. PTAL

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc (hunk 2 above)
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:229]: Added checking size in this function, PTAL.

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 3]
```diff
@@ -221,17 +227,18 @@ snpe_subplugin::runtimeToString (zdl::DlSystem::Runtime_t runtime)
...
 {
+  if (output_tensor_names_list.size () > 0) {
+    output_tensor_names_list = zdl::DlSystem::StringList ();
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:233]: Isn't this a memory leak? or SNPE manages it?

(the previous contents of output_tensor_names_list is dangled? don't you need ```delete output_tensor_names_list``` ?)

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:233]: This line mean to initialize the list. SNPE manages the destruction.

Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:233]: Ok!

**Review by myungjoo — APPROVED** [review]
LGTM

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 4]
```diff
@@ -221,17 +227,18 @@ snpe_subplugin::runtimeToString (zdl::DlSystem::Runtime_t runtime)
...
+    if (info->info[i].name == nullptr || info->info[i].name[0] == '\0') {
       /* failed */
-      nns_loge ("Given layer name with index %u is invalid.", i);
-      output_layer_names.clear ();
+      nns_loge ("Given output tensor name with index %u is invalid, it is null.", i);
       return false;
     }
-    nns_logd ("Add output layer name of %s", info->info[i].name);
-    output_layer_names.emplace_back (std::string (info->info[i].name));
+    output_tensor_names_list.append (g_strdup (info->info[i].name));
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:241]: line 241 and 312
does it need memcpy - strdup()?

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc]: Thanks! It was memleak. I updated the code

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 5]
```diff
@@ -290,19 +297,21 @@ snpe_subplugin::parse_custom_prop (const char *custom_prop)
...
-            nns_loge ("Given layer name with index %u is invalid.", i);
-            output_layer_names.clear ();
+            nns_loge ("Given tensor name with index %u is invalid.", i);
             invalid_option = true;
             break;
           }
-          nns_logd ("Add output layer name of %s", names[i]);
-          output_layer_names.emplace_back (std::string (names[i]));
+          nns_logd ("Add output tensor name of %s", names[i]);
+          output_tensor_names_list.append (g_strdup (names[i]));
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:312]: plz fix this line also.

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc]: Thanks! PTAL

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**